### PR TITLE
Report error for invalid 'this' type during 'await'

### LINKED ISSUE #47711: `await` does not respect `this` types in `.then` method
# Bug Report

### 🔎 Search Terms

promise, async, await, then, method, this, type

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Promises and async/await.

### ⏯ Playground Link

[Link](https://www.typescriptlang.org/play?jsx=0&ts=4.6.0-dev.20220116#code/C4TwDgpgBAoglsAFhATgHhgGigQQHxQC8UAMhAGbAYEA+UASnAOaJX4DcAUKJKRVTALEA3lGABDJgC4oAcjKVZ2CDJhQAvlx7RGLNkKiiJ0ubtZKo4mTg1dOAYwD2AOwDOwKAFsA1go-FqAAoVWABKVQRkdCwoZwgAN1QhAkCjSRl5fgtodVCuJzcPHzN-KDR8QKtccNhI1DQ4xJRsfCIUtJNZEotxDTzOTgB6QagASTEUOAgAEyhHTwRgOGcmMWQoAAM4TzAAGwhPCGdgVw2oe3EAV1doKOgAI0uPJAeIRHF4uEcUKGm4Wecjg89neKwgADoHLtxK5XLAAAooeZwG4YFoEbZ7A5HE5QRHIm4kODeCDlAjCThQKlQdziJb2GmXez2CAzMmVaw1GD4haoxqodGGSnUkUoCDAS4oZyxCAAdwRSN5EECPJRELFrkcu0SgWKzFYlVCRq4IvUAxFtPpUHI4jguyCIRgXNVqJi-JQ5OFIqpYolUpl8u5irVKuDN3BGq1Op8fmCRv6pvN1IK7hQTOA30CYvE0xcuxAUDAMhdECJJIwdWi6LwoUMGiTVJezjQACEiLhsC2AIzt914QJe73DBUE6AXaXieweGGWKAAFWQznE932c2c+agssXmxgZxRm3dG0h3upSBRqhLDQSAtweEwg5FLnIl125Dt+2mAH4ZIFAvFxLslwhDgtaEAQbZ0CWZaki2NZQHQzgvrs8FQJczjTBQywzPeJ5Ui4YoAFYQFOMzflAv7ZpqzgyOIzggKB4E9pBYalsSMFdnBCFIShaEYW+cTTIONRQWxrYod2nq4VAvqStKZ6uOCYDgk2A5SSKw4AMIwksKxiI4mwlGSe5wq44jkLcqCvBc1zQLKECyIkUAQLs2zLHSMxrNAYCOLCcD3HaCAFo45CbH41BHg+uEQJWbRrs+r7vqR4K-tFLw-DOhn4KE4LiLWn6fuRqVRJYcKZTWOUlahzjeICsoTnCLY4WpeHOIRxHADMkVUqEg5mmaDguO4m6RHO4AzAAYragFiheLENJcnj3DeqbLEwBhBqO4I2nagRdv0QwjDAKBIigcIzhAAAekAkdMUgHQdUAAEQLtAsjybI5wuB1F0eMFYhjXIG1KvNi3LcAkwrHgH37oC06+UwS4rtAGZeOKiCONMsiuA9IpvYgKIfX92iA5e7rYCtkOPZwsojWN0yTXakoQipcJgTSymOAAqmAkAoFpNyBEaymLoEKZahCuyOEw+0wiAzgMs+8tLC4YgQO4gtCupIwACKOGrsRAoWSLTEy0C0U5ACOlxwP++zHE5x3fMe1LDgA6vZYqWMAHU7Dpqwo5OLKwkbji86AmziGcf29BshkU2tZyPB4yzSRAAGBZutwfGboX8CDS0ehsTVUsOhy0atax0psqZnHcUD7hsvGYQJEUiimHipu24iyraHg00go2QPTU1M3YWvzvjcI07syG1ecVw3LOabHNs0CoCdzs+uKsk0uDHPc7z-PKv0-UderPUisLRyi4N4vgpL0uDuCFzACCN9uHf6-fLWw4ANowPQegABdGQWlnCw1TjmEOYcphwj+k3fiHksxpz+LpN6XMeaoCPrIUIQA)

### 💻 Code

```ts
type Either<E, A> = Left<E> | Right<A>;
type Left<E> = { tag: 'Left', e: E };
type Right<A> = { tag: 'Right', a: A };

const mkLeft = <E>(e: E): Either<E, never> => ({ tag: 'Left', e });
const mkRight = <A>(a: A): Either<never, A> => ({ tag: 'Right', a });

// I tried omitting the `implements` cause here but the behavior did not change.
class EPromise<E, A> implements PromiseLike<A> {
    static succeed<A>(a: A): EPromise<never, A> {
        return new EPromise(Promise.resolve(mkRight(a)));
    }

    static fail<E>(e: E): EPromise<E, never> {
        return new EPromise(Promise.resolve(mkLeft(e)));
    }

    constructor(readonly p: PromiseLike<Either<E, A>>) { }

    then<B = A, B1 = never>(
        // EPromise can act as a Thenable only when `E` is `never`.
        this: EPromise<never, A>,
        onfulfilled?: ((value: A) => B | PromiseLike<B>) | null | undefined,
        onrejected?: ((reason: any) => B1 | PromiseLike<B1>) | null | undefined
    ): PromiseLike<B | B1> {
        return this.p.then(
            // Casting to `Right<A>` is safe here because we've eliminated the possibility of `Left<E>`.
            either => onfulfilled?.((either as Right<A>).a) ?? (either as Right<A>).a as unknown as B,
            onrejected
        )
    }
}

const withTypedFailure: EPromise<number, string> = EPromise.fail(1);

// Errors as expected:
//
// "The 'this' context of type 'EPromise<number, string>' is not assignable to method's
//     'this' of type 'EPromise<never, string>"
withTypedFailure.then(s => s.toUpperCase()).then(console.log);

async function test() {
    // Does not produce an equivalent error.
    // We're attempting to access property `a` of a `Right<string>` but in reality we have a `Left<number>`,
    // meaning that `str` here is `undefined`.
    const str = await withTypedFailure;

    // This will now cause a runtime error.
    return str.toUpperCase();
}

test()
    .then(console.log)
    .catch(console.error) // [ERR]: Cannot read properties of undefined (reading 'toUpperCase')
```

### 🙁 Actual behavior

Awaiting a `Thenable` with a user-defined `this` type in the `.then` method does not respect the invariants imposed by the `this` type.

### 🙂 Expected behavior

Awaiting a `Thenable` with a user-defined `this` type in the `.then` method should respect the invariants imposed by the `this` type and produce the same error as making an explicit call to `.then`.

### PR DESCRIPTION
**Pull request by rbuckton**
Fixes #47711

### CODE REVIEW

**Review by sandersn — APPROVED** [review]
Looks good, although I would personally collapse the two error parameters into one since they seem to be mutually exclusive.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -36256,7 +36256,7 @@ namespace ts {
          * @param type The type of the promise.
          * @remarks The "promised type" of a type is the type of the "value" parameter of the "onfulfilled" callback.
          */
-        function getPromisedTypeOfPromise(type: Type, errorNode?: Node): Type | undefined {
+        function getPromisedTypeOfPromise(type: Type, errorNode?: Node, thisTypeForErrorOut?: { value?: Type }): Type | undefined {
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/compiler/checker.ts:36259]: I don't want to have to test whether `errorNode` is a `Node` and I'd also rather not allocate a nursery object for `{ errorNode }` if I don't have to.

**Comment by rbuckton** [comment]
@DanielRosenwasser worth taking for 4.7? If so I can merge now.

**Comment by rbuckton** [comment]
@typescript-bot test this
@typescript-bot user test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot user test this inline
@typescript-bot perf test faster

**Pull request merged**